src/conn_poll.c

```diff
@@ -141,299 +313,487 @@ int conn_poll_prepare(conn_registry_t *registry, pfds_record_t *pfds, timestamp_
+	}
+}
 
-		if (ret == -SEAGAIN || ret == -SEWOULDBLOCK) {
-			JLOG_VERBOSE("No more ICE-TCP datagrams to receive");
-		} else if (ret <= 0) {
-			if (ret == 0) JLOG_DEBUG("TCP connection closed");
-			else JLOG_DEBUG("TCP connection failed");
-			conn_poll_change_tcp_fail(agent);
-			return;
-		}
+static int conn_poll_tcp_finish_connect(juice_agent_t *agent, tcp_connection_t *connection) {
+	int error = 0;
+	socklen_t len = sizeof(error);
+	if (getsockopt(connection->sock, SOL_SOCKET, SO_ERROR, (void *)&error, &len) < 0) {
+		JLOG_WARN("getsockopt(SO_ERROR) failed, errno=%d", sockerrno);
+		return -1;
+	}
+	if (error != 0) {
+		JLOG_WARN("ICE-TCP connect failed, error=%d", error);
+		return -1;
+	}
+	connection->connecting = false;
+	if (agent_conn_tcp_state(agent, &connection->remote, TCP_STATE_CONNECTED) != 0)
+		return -1;
+	return 0;
+}
 
-		if (agent_conn_update(agent, &conn_impl->next_timestamp) != 0) {
-			JLOG_WARN("Agent update failed");
-			conn_impl->state = CONN_STATE_FINISHED;
-			return;
-		}
+static tcp_connection_t *conn_poll_tcp_create_active(juice_agent_t *agent, conn_impl_t *conn_impl,
+                                                    const addr_record_t *dst) {
+	if (conn_poll_tcp_reserve(conn_impl, conn_impl->tcp_conns_count + 1))
+		return NULL;
 
-	} else if (conn_impl->next_timestamp <= current_timestamp()) {
-		if (agent_conn_update(agent, &conn_impl->next_timestamp) != 0) {
-			JLOG_WARN("Agent update failed");
-			conn_impl->state = CONN_STATE_FINISHED;
-			return;
+	socket_t sock = socket(dst->addr.ss_family, SOCK_STREAM, IPPROTO_TCP);
+	if (sock == INVALID_SOCKET) {
+		JLOG_WARN("ICE-TCP socket creation failed, errno=%d", sockerrno);
+		return NULL;
+	}
+
+	const sockopt_t buffer_size = 1 * 1024 * 1024;
+	setsockopt(sock, SOL_SOCKET, SO_RCVBUF, (const char *)&buffer_size, sizeof(buffer_size));
+	setsockopt(sock, SOL_SOCKET, SO_SNDBUF, (const char *)&buffer_size, sizeof(buffer_size));
+
+	ctl_t nbio = 1;
+	if (ioctlsocket(sock, FIONBIO, &nbio)) {
+		JLOG_WARN("ICE-TCP set non-blocking failed, errno=%d", sockerrno);
+		closesocket(sock);
+		return NULL;
+	}
+
+	int ret = connect(sock, (const struct sockaddr *)&dst->addr, dst->len);
+	bool connecting = false;
+	if (ret < 0) {
+		if (sockerrno == SEINPROGRESS || sockerrno == SEWOULDBLOCK) {
+			connecting = true;
+		} else {
+			JLOG_WARN("ICE-TCP connect failed immediately, errno=%d", sockerrno);
+			closesocket(sock);
+			return NULL;
 		}
 	}
-}
 
-void conn_poll_change_tcp_fail(juice_agent_t *agent) {
-	conn_impl_t *conn_impl = agent->conn_impl;
-	if (conn_impl->tcp_sock != INVALID_SOCKET) {
-		closesocket(conn_impl->tcp_sock);
-		conn_impl->tcp_sock = INVALID_SOCKET;
+	tcp_connection_t *connection = conn_impl->tcp_conns + conn_impl->tcp_conns_count;
+	memset(connection, 0, sizeof(*connection));
+	connection->sock = sock;
+	connection->remote = *dst;
+	connection->remote.socktype = SOCK_STREAM;
+	connection->closed = false;
+	connection->connecting = connecting;
```

Comment:
`connecting` is always true here. Also, why use two flags when there is an already-defined state enum?